pkg/image/cdi/upload.go

```diff
@@ -1,20 +1,334 @@
+
+	dataContent := rawContent
+	if multipartFormat {
+		dataContent = rawContent[headerEndIndex+4:]
+	}
+	if _, err := cu.vmio.UpdateVirtualSizeAndSize(vmImg, virtualSize, fileSize); err != nil {
+		return fmt.Errorf("failed to update VM Image size and virtual size: %v", err)
+	}
+
+	// check VMImage status again (for size/virtual size)
+	if err := wait.PollUntilContextTimeout(context.Background(), tickPolling, tickTimeout, true, func(context.Context) (bool, error) {
+		return cu.waitVMImageStatus(vmImg, fileSize, virtualSize)
+	}); err != nil {
+		return fmt.Errorf("failed to wait for VMImage status: %v", err)
+	}
+
+	// get the latest VMImage
+	vmImg, err = cu.vmio.GetVMImageObj(vmImg.Namespace, vmImg.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get VMImage: %v", err)
+	}
+
+	// create DataVolume
+	dvName := cu.vmio.GetName(vmImg)
+	dvNamespace := cu.vmio.GetNamespace(vmImg)
+
+	logrus.Infof("The VM Image Status updated, start to create DataVolume %s/%s", dvNamespace, dvName)
+
+	// generate DV source
+	dvSource, err := generateDVSource(vmImg, cu.vmio.GetSourceType(vmImg))
+	if err != nil {
+		return fmt.Errorf("failed to generate DV source: %v", err)
+	}
+
+	// generate DV target storage
+	dvTargetStorage, err := generateDVTargetStorage(vmImg)
+	if err != nil {
+		return fmt.Errorf("failed to generate DV target storage: %v", err)
+	}
+	var boolTrue = true
+	dataVolumeTemplate := &cdiv1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dvName,
+			Namespace: dvNamespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         common.HarvesterAPIV1Beta1,
+					Kind:               common.VMImageKind,
+					Name:               cu.vmio.GetName(vmImg),
+					UID:                cu.vmio.GetUID(vmImg),
+					BlockOwnerDeletion: &boolTrue,
+				},
+			},
+		},
+		Spec: cdiv1.DataVolumeSpec{
+			Source:  dvSource,
+			Storage: dvTargetStorage,
+		},
+	}
+	if _, err := cu.dataVolumeClient.Create(dataVolumeTemplate); err != nil {
+		return fmt.Errorf("failed to create DataVolume %s/%s: %v", dvNamespace, dvName, err)
+	}
+	logrus.Infof("DataVolume %s/%s created", dvNamespace, dvName)
+
+	// wait data volume UploadReady
+	if err := wait.PollUntilContextTimeout(context.Background(), tickPolling, tickTimeout, true, func(context.Context) (bool, error) {
+		return cu.waitDataVolumeStatus(dvNamespace, dvName, cdiv1.UploadReady)
+	}); err != nil {
+		return fmt.Errorf("failed to wait for VMImage status: %v", err)
+	}
+
+	uploadTokenRequest := &uploadcdiv1.UploadTokenRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "token-for-upload",
+			Namespace: dvNamespace,
+		},
+		Spec: uploadcdiv1.UploadTokenRequestSpec{
+			PvcName: dvName,
+		},
+	}
+	retUploadTokenRequest, err := cu.cdiUploadClient.Create(uploadTokenRequest)
+	if err != nil {
+		return fmt.Errorf("failed to create UploadTokenRequest %s/%s: %v", dvNamespace, dvName, err)
+	}
+
+	newBody := io.MultiReader(bytes.NewReader(dataContent), req.Body)
```

Comment:
IIUC, `dataContent` is the first N bytes of `req.Body` (multipart or not). Won't this `io.MultiReader()` end up reading redundant data from `req.Body`?

Reply:
No, it will wrap these two contents again for later io reader.